### PR TITLE
[Rough Sketch] Normalize vector output

### PR DESCRIPTION
**Pull request by NegatioN**
Hi!
Normalized outputs are often easier to reason about, and threshold for recommendation systems, and I've observed in other models that training with normalization included can improve precision. (compared to normalizing afterwards)

Is this anything you've considered in lightfm? Is there any obvious reason as to why this shouldn't be here? (not compatible with the way individual vectors are being added together parhaps?)

With this PR, all previous code works as before, but when running unmodified tests with `normalize=True` I did get infinities in parameters, and I'm sadly not proficient enough in cython to debug this (yet).

However if this looks like a breeze to implement, or you could really see this as an improvement, maybe that could be fixed and tested somehow.

Any thoughts?

### CODE REVIEW

File: lightfm/lightfm.py [hunk 1]
```diff
@@ -94,6 +94,10 @@ class LightFM(object):
     random_state: int seed, RandomState instance, or None
         The seed of the pseudo random number generator to use when shuffling
         the data and initializing the parameters.
+    normalize: bool, optional
```
Comment (maciejkula) [inline lightfm/lightfm.py]: Can we reformulate this to better reflect the fact that this affects training as well? Something like the following:

If True, user and item vectors will be rescaled to unit length during training and prediction.

File: lightfm/_lightfm_fast.pyx.template [hunk 2]
```diff
@@ -316,10 +317,20 @@ cdef inline void compute_representation(CSRMatrix features,
 
         representation[lightfm.no_components] += feature_weight * feature_biases[feature]
 
+cdef inline flt normalize_repr(flt *v_repr, int no_components) nogil:
```
Comment (maciejkula) [inline lightfm/_lightfm_fast.pyx.template]: Can we call this `compute_norm`? I think it better reflects what this is actually doing.

**Comment by NegatioN** [comment]
Sorry for the late reply :) Cool that you like it, I'll try to write some tests during next week, and hopefully results will make enough sense that we can feel comfortable merging it.

**Comment by maciejkula** [comment]
Sounds good!

**Comment by NegatioN** [comment]
I think there are some aspects of the code that I don't fully understand, and that are not immediately compatible with normalized vectors. 
Do you have any tips/pointers for what might be causing this?

When i try to run a simple movielens-test on the data, `loss=logistic` item_embeddings diverges, and `loss=warp` gets poor performance on both test an train data.
I will continue to investigate
```
============================================================================================================== FAILURES ==============================================================================================================
_______________________________________________________________________________________________ test_movielens_normalized_accuracy_fit _______________________________________________________________________________________________

    def test_movielens_normalized_accuracy_fit():
        model = LightFM(random_state=SEED, normalize=True, loss='warp')
        model.fit(train, epochs=20)

        train_predictions = model.predict(train.row,
                                          train.col)
        test_predictions = model.predict(test.row,
                                         test.col)

        #assert roc_auc_score(train.data, train_predictions) > 0.84
>       assert roc_auc_score(test.data, test_predictions) > 0.76
E       assert 0.63457222732392338 > 0.76
E        +  where 0.63457222732392338 = roc_auc_score(array([ 1,  1,  1, ..., -1,  1, -1], dtype=int32), array([-0.51581025,  0.03291166, -0.27147731, ..., -0.23394495,\n       -0.53144515, -0.6818071 ]))
E        +    where array([ 1,  1,  1, ..., -1,  1, -1], dtype=int32) = <943x1682 sparse matrix of type '<type 'numpy.int32'>'\n	with 9430 stored elements in COOrdinate format>.data
```

**Comment by NegatioN** [comment]
I think this has to do with the addition of multiple normalized vectors onto each other. This doesn't seem to gel well with stable results. Don't really have the time to look for a solution now, but it would be cool if anyone else dropping by this PR looked into it :)

**Comment by maciejkula** [comment]
Have you tried normalizing after the weighted summation?

**Comment by NegatioN** [comment]
I didn't write the correctly in my last post, this is what I meant:
I don't think the normalization works well after the weighted summation step.

As far as I understand this should be after summation and almost the last step before calculating the loss. Code for normalizing looked like:
```
cdef inline flt compute_norm(flt *v_repr, int no_components) nogil:

    cdef int i
    cdef flt result

    for i in range(no_components):
        result += v_repr[i] * v_repr[i]
    return sqrt(result)


cdef inline flt compute_prediction_from_repr(flt *user_repr,
                                             flt *item_repr,
                                             int no_components,
                                             bint normalize) nogil:

    cdef int i
    cdef flt result

    # Biases
    result = user_repr[no_components] + item_repr[no_components]

    # Latent factor dot product
    for i in range(no_components):
        result += user_repr[i] * item_repr[i]

    if normalize:
        result = result / (compute_norm(user_repr, no_components) * compute_norm(item_repr, no_components))

    return result
```

However, I now tried normalizing at the summation-step... for science or something, and that didn't seem to produce any different results, so maybe I'm doing something fundamentally wrong somewhere.

```
cdef inline void compute_representation(CSRMatrix features,
                                        flt[:, ::1] feature_embeddings,
                                        flt[::1] feature_biases,
                                        FastLightFM lightfm,
                                        int row_id,
                                        double scale,
                                        flt *representation) nogil:
    """
    Compute latent representation for row_id.
    The last element of the representation is the bias.
    """

    cdef int i, j, start_index, stop_index, feature
    cdef flt feature_weight
    cdef flt norm
    cdef flt *current_emb = <flt *> malloc(lightfm.no_components+1 * sizeof(flt))

    start_index = features.get_row_start(row_id)
    stop_index = features.get_row_end(row_id)

    for i in range(lightfm.no_components + 1):
        representation[i] = 0.0

    for i in range(start_index, stop_index):
        feature = features.indices[i]
        for j in range(lightfm.no_components+1):
            current_emb[j] = feature_embeddings[feature, j]
        norm = 1 / compute_norm(current_emb, lightfm.no_components+1)
        feature_weight = features.data[i] * scale * norm
        for j in range(lightfm.no_components):
            representation[j] += feature_embeddings[feature, j] * feature_weight

        representation[lightfm.no_components] += feature_biases[feature] * feature_weight

    free(current_emb)
```
